futures-util/src/stream/futures_unordered/node.rs

```diff
@@ -0,0 +1,145 @@
+
+    // Previous node in linked list tracking all active nodes
+    pub(super) prev_all: UnsafeCell<*const Node<T>>,
+
+    // Next pointer in readiness queue
+    pub(super) next_readiness: AtomicPtr<Node<T>>,
+
+    // Queue that we'll be enqueued to when notified
+    pub(super) queue: Weak<Inner<T>>,
+
+    // Whether or not this node is currently in the mpsc queue.
+    pub(super) queued: AtomicBool,
+}
+
+impl<T> Node<T> {
+    pub(super) fn notify(me: &Arc<Node<T>>) {
+        let inner = match me.queue.upgrade() {
+            Some(inner) => inner,
+            None => return,
+        };
+
+        // It's our job to notify the node that it's ready to get polled,
+        // meaning that we need to enqueue it into the readiness queue. To
+        // do this we flag that we're ready to be queued, and if successful
+        // we then do the literal queueing operation, ensuring that we're
+        // only queued once.
+        //
+        // Once the node is inserted we be sure to notify the parent task,
+        // as it'll want to come along and pick up our node now.
+        //
+        // Note that we don't change the reference count of the node here,
+        // we're just enqueueing the raw pointer. The `FuturesUnordered`
+        // implementation guarantees that if we set the `queued` flag true that
+        // there's a reference count held by the main `FuturesUnordered` queue
+        // still.
+        let prev = me.queued.swap(true, SeqCst);
+        if !prev {
+            inner.enqueue(&**me);
+            inner.parent.wake();
+        }
+    }
+}
+
+impl<T> Drop for Node<T> {
+    fn drop(&mut self) {
+        // Currently a `Node<T>` is sent across all threads for any lifetime,
+        // regardless of `T`. This means that for memory safety we can't
+        // actually touch `T` at any time except when we have a reference to the
+        // `FuturesUnordered` itself.
+        //
+        // Consequently it *should* be the case that we always drop futures from
+        // the `FuturesUnordered` instance, but this is a bomb in place to catch
+        // any bugs in that logic.
+        unsafe {
+            if (*self.future.get()).is_some() {
+                abort("future still here when dropping");
+            }
+        }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub(super) struct NodeToHandle<'a, T: 'a>(pub(super) &'a Arc<Node<T>>);
+
+impl<'a, T> Clone for NodeToHandle<'a, T> {
+    fn clone(&self) -> Self {
+        NodeToHandle(self.0)
+    }
+}
+
+#[doc(hidden)]
+impl<'a, T> From<NodeToHandle<'a, T>> for LocalWaker {
+    fn from(handle: NodeToHandle<'a, T>) -> LocalWaker {
+        unsafe {
+            let ptr: Arc<Node<T>> = handle.0.clone();
+            let ptr = mem::transmute::<Arc<Node<T>>, *mut ArcNode<T>>(ptr);
+            let ptr = ptr as *mut UnsafeWake;
+            // Hide lifetime
+            let ptr = mem::transmute::<*mut UnsafeWake, *mut UnsafeWake>(ptr);
```

Comment:
This looks like a no-op?

Reply:
It hides the lifetime. Before my PR it was in a function called `hide_lt`. Otherwise we get:
```
error[E0310]: the parameter type `T` may not live long enough
  --> futures-util/src/stream/futures_unordered/node.rs:96:23
   |
91 | impl<'a, T> From<NodeToHandle<'a, T>> for LocalWaker {
   |          - help: consider adding an explicit lifetime bound `T: 'static`...
...
96 |             let ptr = ptr as *mut UnsafeWake;
   |                       ^^^
   |
note: ...so that the type `stream::futures_unordered::node::ArcNode<T>` will meet its required lifetime bounds
```

Reply:
Something else is going on here-- the error you're getting is on line 96, not the usage of the ptr. `*mut UnsafeWake` is one type, and it doesn't contain any lifetimes.